base/shared/src/main/scala/scalaz/data/Is.scala

```diff
@@ -0,0 +1,218 @@
+  * may be stored inside a value of type `F[A]`. Hence it is impossible for
+  * a substitution function to alter the value it takes as argument.
+  *
+  * Not taking into account the partial functions that never terminate
+  * (infinite loops), functions returning `null`, or throwing exceptions,
+  * the identity function is the only function that can be used in place of
+  * `subst` to construct a value of type `Is[A, B]`.
+  *
+  * The existence of a value of type `Is[A, B]` now implies that a ≡ b,
+  * since the conversion function, that converts an `A` into a `B`, must be
+  * the identity function.
+  *
+  * This technique was first used in
+  * [[http://portal.acm.org/citation.cfm?id=583852.581494
+  * Typing Dynamic Typing]] (Baars and Swierstra, ICFP 2002).
+  *
+  * @see [[===]] `A === B` is a type synonym to `Is[A, B]`
+  * @see [[http://typelevel.org/blog/2014/09/20/higher_leibniz.html
+  *        Higher Leibniz]]
+  */
+sealed trait Is[A, B] { ab =>
+  import Is._
+
+  /**
+    * To create an instance of `Is[A, B]` you must show that for every
+    * choice of `F[_]` you can convert `F[A]` to `F[B]`.
+    */
+  def subst[F[_]](fa: F[A]): F[B]
+
+  /**
+    * Substitution on identity brings about a direct coercion function of the
+    * same form that `=:=` provides.
+    *
+    * @see [[coerce]]
+    */
+  def apply(a: A): B = coerce(a)
+
+  /**
+    * Substitution on identity brings about a direct coercion function of the
+    * same form that [[=:=]] provides.
+    *
+    * @see [[apply]]
+    */
+  def coerce(a: A): B =
+    subst[λ[α => α]](a)
+
+  /**
+    * Equality is transitive relation and its witnesses can be composed
+    * in a chain much like functions.
+    *
+    * @see [[compose]]
+    */
+  final def andThen[C](bc: B === C): A === C =
+    bc.subst[A === ?](ab)
+
+  /**
+    * Equality is transitive relation and its witnesses can be composed
+    * in a chain much like functions.
+    *
+    * @see [[andThen]]
+    */
+  final def compose[Z](za: Z === A): Z === B =
+    za.andThen(ab)
+
+  /**
+    * Equality is symmetric relation and therefore can be flipped around.
+    * Flipping is its own inverse, so `x.flip.flip == x`.
+    */
+  final def flip: B === A =
+    subst[? === A](refl)
+
+  /**
+    * Given `A === B` we can prove that `F[A] === F[B]`.
+    *
+    * @see [[Is.lift]]
+    * @see [[Is.lift2]]
+    */
+  final def lift[F[_]]: F[A] === F[B] =
+    Is.lift[F, A, B](ab)
+
+  /**
+    * Given `A === B` and `I === J` we can prove that `F[A, I] === F[B, J]`.
+    *
+    * This method allows you to compose two `===` values in infix manner:
+    * {{{
+    *   def either(ab: A === B, ij: I === J): Either[A, I] === Either[B, J] =
+    *     ab lift2[Either] ij
+    * }}}
+    *
+    * @see [[Is.lift]]
+    * @see [[Is.lift2]]
+    * @see [[Is.lift3]]
+    */
+  def lift2[F[_, _]]: PartiallyAppliedLift2[F] =
+    new PartiallyAppliedLift2[F]
+  final class PartiallyAppliedLift2[F[_, _]] {
```

Comment:
You can't do that exactly, because it's a non-static inner class; you could (should) have a `class Is.PartiallyAppliedLift2[A, B] private (val ab: A Is B) extends AnyVal`...